File: api/train_regression.py
from api.python import nn_numpy as nn
import numpy as np
import matplotlib.pyplot as plt


class MyNet(nn.Module):
    def __init__(self, input_dim):
        super().__init__()
        self.input_dim = input_dim
        self.hidden1 = nn.layers.Dense(in_features=self.input_dim,
                                       out_features=10,
                                       activation=nn.acts.Tanh(),
                                       weight_initializer=nn.inits.RandomUniform(),
                                       bias_initializer=nn.inits.Constant(0.)
                                       )

        self.output = nn.layers.Dense(in_features=10,
                                      out_features=1,
                                      weight_initializer=nn.inits.RandomUniform(),
                                      bias_initializer=nn.inits.Constant(0.))

    def forward(self, x):
        x = self.hidden1(x)
        return self.output(x)


np.random.seed(1)
x = np.linspace(-1, 1, 256)[:, None]  # [batch, 1]
t = x ** 2 + np.random.normal(0., 0.1, (256, 1))  # [batch, 1]

my_net = MyNet(1)
mse = nn.losses.MSELoss()
opt = nn.optims.SGD(my_net.parameters, 0.1)
for epoch in range(1000):
    y = my_net(x)
    loss = mse(y, t)
    my_net.backward(loss)
    opt.apply()
    print("Step: %i | loss: %.5f" % (epoch, loss.value))

plt.scatter(x, t, s=20)
plt.plot(x, y, c="red", lw=3)
plt.show()

File: java_nn/src/train_regression.java
import Layers.Dense;
import Losses.Loss;
import Losses.MSELoss;
import java.util.Arrays;
import java.util.Random;
import java.lang.Math;

class MyNet extends Module{
    int in_features = 0;
    Dense hidden1, output;
    public MyNet(int in_features){
        this.in_features = in_features;
        this.hidden1 = new Dense(this.in_features,
                10,
                "relu",
                "random_uniform",
                "zeros");
        this.layers.add(this.hidden1);
        this.output = new Dense(10,
                1,
                "linear",
                "random_uniform",
                "zeros");
        this.layers.add(this.output);
        this.set_params(this.layers);
    }

    public float[][] forward(float[][] x){
        x = this.hidden1.forward(x);
        x = this.output.forward(x);
        return x;
    }
}

public class train_regression {
    public static void main(String[] args) {
        Random random = new Random();
        random.setSeed(1);

        float[][] x = new float[200][1], t = new float[200][1];
        for(int i = -100; i < 100; i++){
            x[i + 100][0] = 0.01F * i;
            t[i + 100][0] = (float) Math.pow(x[i + 100][0], 2) + (float)(random.nextGaussian() * 0.1);
        }
        System.out.println(Arrays.deepToString(x));
        MyNet my_net = new MyNet(1);
        float[][] y = my_net.forward(x);
        System.out.println(Arrays.deepToString(y));
        System.out.println(Arrays.deepToString(t));
        MSELoss mse = new MSELoss();
        Loss loss = mse.apply(y, t);
        System.out.println(loss);




    }
}
